ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamAsyncClusterTests.java

```diff
@@ -69,12 +69,29 @@ public void testMultipleStreamsMultipleServers() throws Exception {
     RaftServerConfigKeys.Rpc.setTimeoutMax(getProperties(), max);
   }
 
+  @Test
+  public void testSingleStreamMultipleServersStepDownLeader() throws Exception {
+    runWithNewCluster(3, this::runTestDataStreamStepDownLeader);
+  }
+
+  void runTestDataStreamStepDownLeader(CLUSTER cluster) throws Exception {
+    runTestDataStream(cluster, true);
+  }
+
   void runTestDataStream(CLUSTER cluster) throws Exception {
+    runTestDataStream(cluster, false);
+  }
+
+  void runTestDataStream(CLUSTER cluster, boolean stepDownLeader) throws Exception {
     RaftTestUtil.waitForLeader(cluster);
 
     final List<CompletableFuture<Long>> futures = new ArrayList<>();
-    futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 5, 10, 1_000_000, 10), executor));
-    futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 2, 20, 1_000, 10_000), executor));
+    if (stepDownLeader) {
+      futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 1, 1, 1_000_000, 10, stepDownLeader), executor));
```

Comment:
Tried to run with more streams but the test may fail once in a while.
```
@@ -87,7 +87,7 @@ public abstract class DataStreamAsyncClusterTests<CLUSTER extends MiniRaftCluste
 
     final List<CompletableFuture<Long>> futures = new ArrayList<>();
     if (stepDownLeader) {
-      futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 1, 1, 1_000_000, 10, stepDownLeader), executor));
+      futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 2, 10, 1_000, 10_000, stepDownLeader), executor));
     } else {
       futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 5, 10, 1_000_000, 10, stepDownLeader), executor));
       futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 2, 20, 1_000, 10_000, stepDownLeader), executor));
```

Reply:
@szetszwo The failed reason seems related with Isolate and deIsolate too many times. The root cause is hard to find. I will continue to find it.

Reply:
I see. Isolate and deIsolate are not designed to run in multiple threads.  We should change leader in a separated thread.

Reply:
We may change leader after the first future has completed.
```
+++ b/ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamAsyncClusterTests.java
@@ -86,17 +86,25 @@ public abstract class DataStreamAsyncClusterTests<CLUSTER extends MiniRaftCluste
     RaftTestUtil.waitForLeader(cluster);
 
     final List<CompletableFuture<Long>> futures = new ArrayList<>();
-    if (stepDownLeader) {
-      futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 1, 1, 1_000_000, 10, stepDownLeader), executor));
-    } else {
-      futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 5, 10, 1_000_000, 10, stepDownLeader), executor));
-      futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 2, 20, 1_000, 10_000, stepDownLeader), executor));
-    }
+    futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 5, 10, 1_000_000, 10, false), executor));
+    futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 2, 20, 1_000, 10_000, false), executor));
     final long maxIndex = futures.stream()
         .map(CompletableFuture::join)
         .max(Long::compareTo)
         .orElseThrow(IllegalStateException::new);
 
+    if (stepDownLeader) {
+      final RaftPeerId oldLeader = cluster.getLeader().getId();
+      final CompletableFuture<RaftPeerId> changeLeader = futures.get(0).thenApplyAsync(dummy -> {
+        try {
+          return RaftTestUtil.changeLeader(cluster, oldLeader);
+        } catch (Exception e) {
+          throw new CompletionException("Failed to change leader from " + oldLeader, e);
+        }
+      });
+      LOG.info("Changed leader from {} to {}", oldLeader, changeLeader.join());
+    }
+
```